build.py

```diff
@@ -0,0 +1,163 @@
+            sys.exit(4)
+
+    if _store_new_version:
+        target = os.path.join(parent_dir, package, version)
+        # Clone the default
+        shutil.copytree(build_dir, target)
+        # Update other variables.
+        build_dir = target
+
+    tentative_yaml_path = os.path.join(build_dir, 'build.yml')
+
+    if not os.path.exists(tentative_yaml_path) and not os.path.isdir(tentative_yaml_path):
+        print "No yaml file found at %s" % tentative_yaml_path
+        sys.exit(3)
+
+    with open(tentative_yaml_path, 'r') as handle:
+        image_data = yaml.load(handle)
+
+    DOCKER_TEMPLATE = """
+FROM %(image)s
+MAINTAINER Nare Coraor <nate@bx.psu.edu>
+
+ENV DEBIAN_FRONTEND noninteractive
+%(meta_env_string)s
+%(env_string)s
+
+VOLUME ["/host"]
+# Pre-build packages
+%(prebuild_packages)s
+
+# Pre-build commands
+%(prebuild_commands)s
+
+ENTRYPOINT ["/bin/bash", "/host/build.sh"]
+    """
+
+    SCRIPT_TEMPLATE = """#!/bin/sh
+urls="
+{url_list}
+"
+
+mkdir -p /build/ && cd /build/;
+
+( for url in $urls; do
+    wget "$url" || false || exit
+done)
+
+{commands}"""
+
+    template_values = {'image': 'ubuntu', 'prebuild_commands': "", 'env_string': ""}
+    # ENVIRONMENT
+    meta_env = {}
+    docker_env = {}
+    # If there's a 'meta' section, copy that into the environment, as we've
+    # stored things like OS/package name/version there historically. Useful now
+    # for the 'default' images which have a version hardcoded.
+    if 'meta' in image_data:
+        meta_env = image_data['meta']
+        template_values['image'] = image_data['meta'].get('image', 'ubuntu')
+
+    # If an image is sepecified, we overwrite the Dockerfile's FROM with that
+    # image
+    if image is not None:
+        template_values['image'] = image
+    # Additionally we store the actual image that the package is built with in
+    # the environment, so it's accessible to build scripts.
+    docker_env['image'] = template_values['image']
+
+    # The docker environment (in the Dockerfile) is built from the image_data's
+    # meta, and 'env' sections.
+    if 'env' in image_data:
+        docker_env.update(image_data['env'])
+    template_values['env_string'] = '\n'.join(['ENV %s %s' % (key, docker_env[key])
+                                               for key in docker_env])
+    template_values['meta_env_string'] = '\n'.join(['ENV %s %s' % (key, meta_env[key])
+                                                    for key in meta_env])
+
+    # Version is often used to construct downloads, if it's set to "default",
+    # then re-set it to the value in the default image's metadat
+    if version == 'default':
+        version = image_data['meta'].get('version', 'default')
+
+    prebuild_packages = ['wget', 'openssl', 'ca-certificates', 'build-essential']
+    if 'prebuild' in image_data and 'packages' in image_data['prebuild']:
+        prebuild_packages.extend(image_data['prebuild']['packages'].strip().split())
```

Comment:
I suppose it's easier for cuttin' and pastin' if the packages are a blob, but being part of structured data it feels somewhat icky that a list is a string.